packages/frontend/app/(dashboard)/course/[cid]/components/QueueChat.tsx

```diff
@@ -0,0 +1,296 @@
+import { Fragment, ReactElement, useEffect, useRef, useState } from 'react'
+import { Alert, Badge, Button, Card, message, Space } from 'antd'
+import { Role } from '@koh/common'
+import UserAvatar from '@/app/components/UserAvatar'
+import { MessageCircleMore } from 'lucide-react'
+import TextArea from 'antd/es/input/TextArea'
+import { API } from '@/app/api'
+import { useQueueChat } from '@/app/hooks/useQueueChat'
+import { CloseOutlined } from '@ant-design/icons'
+
+interface QueueChatProps {
+  role: Role
+  queueId: number
+  questionId: number
+  isMobile: boolean
+  hidden: boolean
+  fixed?: boolean
+  announceNewMessage?: (newCount: number) => void
+  onOpen?: () => void
+  onClose?: () => void
+}
+
+const QueueChat: React.FC<QueueChatProps> = ({
+  role,
+  queueId,
+  questionId,
+  isMobile,
+  hidden,
+  fixed = true,
+  announceNewMessage = (newCount: number) => {
+    return
+  },
+  onOpen = () => {
+    return
+  },
+  onClose = () => {
+    return
+  },
+}): ReactElement => {
+  const [isOpen, setIsOpen] = useState<boolean>(isMobile ? false : true)
+  const [input, setInput] = useState<string>('')
+  const [isLoading, setIsLoading] = useState<boolean>(false)
+  const {
+    queueChatData,
+    queueChatError,
+    mutateQueueChat,
+    newMessageCount,
+    resetNewMessageCount,
+  } = useQueueChat(queueId, questionId)
+  const messagesEndRef = useRef<HTMLDivElement | null>(null) // This handles auto scrolling
+
+  // To always auto scroll to the bottom of the page when new messages are added
+  useEffect(() => {
+    if (messagesEndRef.current && isOpen) {
+      messagesEndRef.current.scrollIntoView({ behavior: 'smooth' })
+    }
+  }, [isOpen, queueChatData, queueChatData?.messages])
+
+  // To handle new message events
+  useEffect(() => {
+    if (newMessageCount == 0 || !queueChatData) {
+      return
+    }
+    if (!isMobile) {
+      // This is for desktop's default behaviour (auto open the chat) -- mobile has css to handle this
+      setIsOpen(true)
+      resetNewMessageCount()
+      onOpen()
+    } else {
+      announceNewMessage(newMessageCount) // For mobile "view chats" button in queue page to know there are new messages
+    }
+  }, [newMessageCount, setIsOpen])
+
+  const isStaff = role === Role.PROFESSOR || role === Role.TA
+
+  const sendMessage = async () => {
+    setIsLoading(true)
+    try {
+      if (questionId) {
+        API.queueChats.sendMessage(queueId, questionId, input).then(() => {
+          mutateQueueChat()
+          setIsLoading(false)
+          setInput('')
+        })
+      }
+    } catch (error) {
+      console.error(error)
+    }
```

Comment:
instead of just printing to console, maybe send a message to the user too. Also setIsLoading(false) should be in finally (or after catch), otherwise it will load forever if there's an error.
```suggestion
      if (questionId) {
        API.queueChats.sendMessage(queueId, questionId, input).then(() => {
          mutateQueueChat()
          setInput('')
        }).catch((e) => {
        message.error(getErrorMessage(e))
        }).finally(() => {
        setIsLoading(false)
        })
      }
```